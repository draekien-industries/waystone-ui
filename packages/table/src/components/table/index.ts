export * from './table.body';
export { TableContainer } from './table.container';
export * from './table.head';
