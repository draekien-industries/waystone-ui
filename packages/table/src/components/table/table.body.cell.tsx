import { flexRender, type Cell } from '@tanstack/react-table';

export const TableBodyCell = <D, V>({ id, column, getContext }: Cell<D, V>) => {
  const { columnDef, getSize } = column;

  return (
    <td
      id={id}
      sx={{
        width: getSize(),
        paddingY: 'small',
        paddingX: 'medium',
      }}
    >
      {flexRender(columnDef.cell, getContext())}
    </td>
  );
};

const VirtualizedTableBodyCell = <D, V>({
  id,
  column,
  getContext,
}: Cell<D, V>) => {
  const { columnDef, getSize } = column;

  return (
    <td
      id={id}
      sx={{
        display: 'flex',
        width: getSize(),
        paddingY: 'small',
        paddingX: 'medium',
      }}
    >
      {flexRender(columnDef.cell, getContext())}
    </td>
  );
};

TableBodyCell.Virtualized = VirtualizedTableBodyCell;
