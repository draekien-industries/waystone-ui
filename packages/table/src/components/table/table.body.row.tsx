import type { Row } from '@tanstack/react-table';
import { alpha } from '@theme-ui/color';
import { TableBodyCell } from './table.body.cell';

export const TableBodyRow = <D,>({
  id,
  getCanSelect,
  getVisibleCells,
}: Row<D>) => {
  const enabled = getCanSelect();
  const cells = getVisibleCells();

  return (
    <tr
      id={id}
      sx={{
        color: enabled ? 'text' : 'muted',
        borderBottom: '1px solid',
        borderBottomColor: alpha('muted', 0.25),
      }}
    >
      {cells.map((cell) => (
        <TableBodyCell key={cell.id} {...cell} />
      ))}
    </tr>
  );
};
