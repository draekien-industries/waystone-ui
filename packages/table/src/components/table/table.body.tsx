import { alpha } from '@theme-ui/color';
import type { PropsWithChildren, ReactNode } from 'react';
import { TableBodyCell } from './table.body.cell';
import { TableBodyRow } from './table.body.row';

export const TableBody = ({ children }: PropsWithChildren) => (
  <tbody
    sx={{
      'tr:nth-of-type(even)': {
        backgroundColor: alpha('b-400', 0.1),
      },
    }}
  >
    {children}
  </tbody>
);

TableBody.Row = TableBodyRow;
TableBody.Cell = TableBodyCell;

export type VirtualizedTableBodyProps = {
  children: ReactNode;
  getTotalSize: () => number;
};

const VirtualizedTableBody = ({
  children,
  getTotalSize,
}: VirtualizedTableBodyProps) => (
  <tbody
    sx={{
      display: 'grid',
      position: 'relative',
      height: getTotalSize(),
    }}
  >
    {children}
  </tbody>
);

TableBody.Virtualized = VirtualizedTableBody;
