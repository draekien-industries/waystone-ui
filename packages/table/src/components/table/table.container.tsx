import type { PropsWithChildren } from 'react';
import type { ThemeUIStyleObject } from 'theme-ui';

export const style = {
  width: '100%',
  minWidth: '400px',
  boxShadow: 'medium',
  borderCollapse: 'collapse',
} as const satisfies ThemeUIStyleObject;

export const TableContainer = ({ children }: PropsWithChildren) => (
  <table sx={style}>{children}</table>
);
