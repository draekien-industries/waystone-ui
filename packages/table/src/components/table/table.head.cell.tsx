import { flexRender, type Header as HeaderProps } from '@tanstack/react-table';
import type { PropsWithChildren } from 'react';
import { Box } from 'theme-ui';

export const TableHeadCell = <D, V>({
  id,
  colSpan,
  isPlaceholder,
  column,
  children,
  getContext,
  getSize,
}: PropsWithChildren<HeaderProps<D, V>>) => (
  <th
    id={id}
    colSpan={colSpan}
    sx={{
      width: getSize(),
      color: 'b-900',
      paddingY: 'small',
      paddingX: 'medium',
      textAlign: 'left',
      fontWeight: 'semibold',
    }}
  >
    {!isPlaceholder && (
      <Box>
        {flexRender(column.columnDef.header, getContext())}
        {children}
      </Box>
    )}
  </th>
);
