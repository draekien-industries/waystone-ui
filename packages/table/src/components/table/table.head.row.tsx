import type { PropsWithChildren } from 'react';
import type { HeaderGroup as TableHeadRowProps } from '@tanstack/react-table';
import { TableHeadCell } from './table.head.cell';

export const TableHeadRow = <D,>({ id, headers }: TableHeadRowProps<D>) => (
  <tr id={id}>
    {headers.map((header) => (
      <TableHeadCell key={header.id} {...header} />
    ))}
  </tr>
);
