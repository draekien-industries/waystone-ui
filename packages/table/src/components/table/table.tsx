import type { PropsWithChildren } from 'react';

export const WaystoneTable = ({ children }: PropsWithChildren) => (
  <table>{children}</table>
);
