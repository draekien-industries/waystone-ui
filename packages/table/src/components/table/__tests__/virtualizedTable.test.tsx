/* eslint-disable react/jsx-props-no-spreading */
import {
  getCoreRowModel,
  getSortedRowModel,
  useReactTable,
} from '@tanstack/react-table';
import { useVirtualizer } from '@tanstack/react-virtual';
import { render } from '@waystone/testing-library';
import { useRef } from 'react';
import { columns, virtualizableData as data } from '../../../__fixtures__';
import { getMeasureElement } from '../../../utils';
import { TableBody } from '../table.body';
import { TableContainer } from '../table.container';
import { TableHead } from '../table.head';

function VirtualTable() {
  const table = useReactTable({
    data,
    columns,
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
  });

  const { rows } = table.getRowModel();

  // the virtualizer needs to know the scrollable container element
  const tableContainerRef = useRef<HTMLDivElement>(null);

  const rowVirtualizer = useVirtualizer({
    count: rows.length,
    estimateSize: () => 42,
    getScrollElement: () => tableContainerRef.current,
    measureElement: getMeasureElement(),
    overscan: 5,
  });

  return (
    <div
      ref={tableContainerRef}
      style={{
        overflow: 'auto',
        position: 'relative',
        height: 800,
      }}
    >
      <TableContainer.Virtualized>
        <TableHead.Virtualized>
          {table.getHeaderGroups().map((headerGroup) => (
            <TableHead.Row.Virtualized key={headerGroup.id} {...headerGroup} />
          ))}
        </TableHead.Virtualized>
        <TableBody.Virtualized {...rowVirtualizer}>
          {rowVirtualizer.getVirtualItems().map((virtualRow) => {
            const row = rows[virtualRow.index];

            return (
              <TableBody.Row.Virtualized
                key={row.id}
                start={virtualRow.start}
                measureElement={rowVirtualizer.measureElement}
                {...row}
              />
            );
          })}
        </TableBody.Virtualized>
      </TableContainer.Virtualized>
    </div>
  );
}

describe('<Table />', () => {
  it('should match snapshot', () => {
    const { container } = render(<VirtualTable />);

    expect(container).toMatchSnapshot();
  });
});
