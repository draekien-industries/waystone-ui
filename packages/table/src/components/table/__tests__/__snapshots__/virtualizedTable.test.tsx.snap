// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> <TableBodyCell.Virtualized /> should match snapshot 1`] = `
<div>
  <td
    class="css-10r0z04"
    id="1"
  />
</div>
`;

exports[`<Table /> <TableBodyRow.Virtualized /> should match snapshot 1`] = `
<div>
  <tr
    class="css-1da2dbn"
    id="1"
  />
</div>
`;

exports[`<Table /> should match snapshot 1`] = `
<div>
  <div
    style="overflow: auto; position: relative; height: 800px;"
  >
    <table
      class="css-11xsab5"
    >
      <thead
        class="css-xyd1f1"
      >
        <tr
          class="css-15830to"
          id="0"
        >
          <th
            class="css-tlfye8"
            id="table-interactions"
          >
            <div
              class="css-1yz7e9k"
            >
              <div
                class="css-100ak8h"
              >
                <div
                  class="css-15cclx7"
                >
                  <input
                    class="css-kzrhrz"
                    id="select-all-checkbox"
                    type="checkbox"
                  />
                </div>
              </div>
            </div>
          </th>
          <th
            class="css-tlfye8"
            id="firstName"
          >
            <div
              class="css-1yz7e9k"
            >
              First Name
            </div>
          </th>
          <th
            class="css-tlfye8"
            id="lastName"
          >
            <div
              class="css-1yz7e9k"
            >
              Last Name
            </div>
          </th>
          <th
            class="css-tlfye8"
            id="age"
          >
            <div
              class="css-1yz7e9k"
            >
              Age
            </div>
          </th>
          <th
            class="css-tlfye8"
            id="visits"
          >
            <div
              class="css-1yz7e9k"
            >
              Visits
            </div>
          </th>
          <th
            class="css-tlfye8"
            id="status"
          >
            <div
              class="css-1yz7e9k"
            >
              Status
            </div>
          </th>
          <th
            class="css-tlfye8"
            id="progress"
          >
            <div
              class="css-1yz7e9k"
            >
              Progress
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class="css-icyf4f"
      />
    </table>
  </div>
</div>
`;
